### PR TITLE
sql: infect more things with sqlbase.TableDescriptor interface

### PR DESCRIPTION
**Pull request by ajwerner**
This works to adopt the interface rather than concrete implementation of
TableDescriptor inside of pg_catalog and crdb_internal. Also picks it up
in the virtualTabler and cleans up LookupTableByID.

Release note: None

### CODE REVIEW

**Comment by ajwerner** [comment]
I started doing it and then it was just off to the races.

**Comment by ajwerner** [comment]
Hold off on looking at this for a minute. Sorry to say that it too is growing.

**Comment by ajwerner** [comment]
This is RFAL

File: pkg/sql/sqlbase/structured.go
Comment (rohany) [inline pkg/sql/sqlbase/structured.go:92]: Why not just return the whole primary index here?

File: pkg/sql/sqlbase/structured.go
Comment (rohany) [inline pkg/sql/sqlbase/structured.go:631]: Is there a reason to keep this function now?

**Review by rohany — COMMENTED** [review]
I think we should try to adopt similar interfaces as used in the optimizer's catalog -- have you looked at those?

**Review by ajwerner — COMMENTED** [review]
Alright this is rebased. I'll get the next one queued up behind it. 

> I think we should try to adopt similar interfaces as used in the optimizer's catalog -- have you looked at those?

all in due time

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52497#-:-MF7bHu9Ai4R-vUnJk3J:bjukybg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @lucy-zhang, and @rohany)

___
*[pkg/sql/sqlbase/structured.go, line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52497#-ME4xOGkQtVDH0R4j6OK-r1-92:-MF7bT1KEGRcV4nviFKf:b-z6bk0j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c838e383203b294751aeae91613acba5147bbc8c/pkg/sql/sqlbase/structured.go#L92)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Why not just return the whole primary index here?
</blockquote></details>

I don't have a good answer here. I was in a mood that `desc.GetPrimaryIndex().ID` was worse but I don't have any principled argument. I'll happily move it back after this all gets merged. 
___
*[pkg/sql/sqlbase/structured.go, line 631 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/52497#-ME4xOGkQtVDH0R4j6OK-r1-631:-MF7bLpn6BdLjcaxugPi:b-skfatr) ([raw file](https://github.com/cockroachdb/cockroach/blob/c838e383203b294751aeae91613acba5147bbc8c/pkg/sql/sqlbase/structured.go#L631)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Is there a reason to keep this function now?
</blockquote></details>

Not really, it's just a shorthand but I'd prefer to just carry it along for a bit until code has been moved around more if that's okay with you.

**Review by rohany — APPROVED** [review]

**Review by thoszhang — APPROVED** [review]

**Comment by ajwerner** [comment]
TFTR!

bors r+

**Comment by otan** [comment]
i don't think bors is automatically retrying

bors r=lucy-zhang

**Comment by otan** [comment]
bors r=lucy-zhang

**Pull request merged**
